src/plugins/data/public/search/aggs/agg_config.ts

```diff
@@ -281,7 +294,61 @@ export class AggConfig {
       enabled: this.enabled,
       type: this.type && this.type.name,
       schema: this.schema,
-      params: outParams,
+      params: outParams as SerializableState,
+    };
+  }
+
+  /**
+   * @deprecated - Use serialize() instead.
+   */
+  toJSON(): AggConfigSerialized {
+    return this.serialize();
+  }
+
+  /**
+   * @returns Returns an ExpressionAst representing the function for this agg type.
+   */
+  toExpressionAst(): ExpressionAstExpression | undefined {
+    const functionName = this.type && this.type.expressionName;
+    const { type, ...rest } = this.serialize();
+    if (!functionName || !rest.params) {
+      // Return undefined - there is no matching expression function for this agg
+      return;
+    }
+
+    // Go through each of the params and convert to an array of expression args.
+    const params = Object.entries(rest.params).reduce((acc, [key, value]) => {
+      const deserializedParam = this.getAggParams().find(p => p.name === key);
+
+      if (deserializedParam && deserializedParam.type === 'agg') {
+        // We can call `toExpressionAst` if our param is another agg
+        acc[key] = [this.getParam(key).toExpressionAst()];
+      } else if (typeof value === 'object') {
+        // For some params which are object, like ranges, we stringify for now
+        acc[key] = [JSON.stringify(value)];
```

Comment:
Making a big assumption for now that each agg param which is an object would be stringified and parsed by the relevant agg type function. Take for example `ranges` in the `range` agg:
```
          ranges: [
            { from: 0, to: 1000 },
            { from: 1000, to: 2000 },
          ]
```

@ppisljar and I discussed a possible alternative of extending the param types to include `toExpressionAst()` as well, so that every param has to tell you how to turn itself into AST. But this would assume that every object param would also provide its own function to use in a subexpression.

This feels like a problem we could probably solve later IMO, as long as the `agg*` functions are all careful to parse these strings. The other caveat is that the param interfaces defined in each of the agg types would be specifying `string` for each of these, instead of their true shape. Ideally I'd like for these to match, so we'll need to be clever about how these are handled, perhaps overriding them to be strings in the interface for the expression function?

Reply:
> @ppisljar and I discussed a possible alternative of extending the param types to include toExpressionAst() as well, so that every param has to tell you how to turn itself into AST. But this would assume that every object param would also provide its own function to use in a subexpression.

We revisited this idea again, and came upon a solution where we can progressively move each object-based param type over to having its own expression function without doing everything at once.

The plan is:

- We add `param.toExpressionAst(value)` as an optional method in each of affected the param types, including the `agg` param type. 
- These params would return an `ExpressionAstFunction` based on the provided values
- In `AggConfig.toExpressionAst()` we simplify the logic to simply check if that method exists on the param, and call it if it does, passing in the value.
- If `param.toExpressionAst()` is not found, we fall back to `JSON.stringify(value)`
- Later, we can gradually add `toExpressionAst` functions as we create dedicated functions for the object-based param types

This will also solve the issue of `AggConfig.toExpressionAst()` needing to return `ExpressionAstExpression` instead of the preferred `ExpressionAstFunction`. By placing `toExpressionAst` on the param level, we can wrap the returned function inside of `ExpressionAstExpression`, since we know that any params which provide expression functions would by necessity be used as a subexpression.